feat(appset): Policies create-only, create-update, sync per ApplicationSet

speedfl opened the pull request:
Currently `create-only`, `create-update`, `sync` policies are only part of the controller launch parameters which does not allow to selectively disable deletion/updates.

## Motivations

- Provide a way to temporarily to deactivate the Applications update (ie: Disable AutoSync on one of the generated Applications) without impacting all projects/teams

- Provide a way to deactivate Applications deletion on a set of ApplicationSet and enable it on other without several instances of the ApplicationSet controller

Signed-off-by: gmuselli <geoffrey.muselli@gmail.com>

Closes #11073

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Note: I am not able to remove the kubebuilder annotations: https://github.com/go-swagger/go-swagger/issues/2687

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [X] The title of the PR states what changed and the related issues number (used for the release note).
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [X] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [X] Does this PR require documentation updates?
* [X] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

speedfl commented:
@crenshaw-dev could you please help me on this ? I am not sure to understand what should I do. 
I guess there is something related to the [rbac configuration](https://argo-cd.readthedocs.io/en/stable/operator-manual/rbac/) but I don't know where and how should I perform my changes. If you maybe have some examples

crenshaw-dev commented:
@speedfl you're asking for clarification in reference to my suggested change in policy override behavior? Or is your question regarding something else?

speedfl commented:
@crenshaw-dev sorry it was not really clear.
I was asking for clarification in reference to your suggested change in policy override behavior.

speedfl commented:
@crenshaw-dev I added the `--allow-policy-override` which is disabled by default if `--policy` is specified.

Added as well unit tests.
Integration tests are ok Locally

crenshaw-dev commented:
@speedfl can you rebase?

crenshaw-dev commented on docs/operator-manual/applicationset/Controlling-Resource-Modification.md:
```diff
@@ -14,17 +14,49 @@ To enable dry-run, add `--dryrun true` to the ApplicationSet Deployment's contai
+
+If the controller parameter `--policy` is set, it takes precedence on the field `applicationsSync`. It is possible to allow per ApplicationSet sync policy by setting controller parameter `--allow-policy-override` (default to `false`).
+
+This does not prevent deletion of Applications if the ApplicationSet is deleted
```
I need to read the PR more, but why can't `syncPolicy.applicationsSync=create-only` prevent Application deletion?

speedfl replied:
It prevents the deletion in case the result of the generator removes an application. But from what I saw in the code the deletion of the applicationset deletes the application:

https://github.com/argoproj/argo-cd/blob/master/applicationset/controllers/applicationset_controller.go#L195

speedfl commented:
@crenshaw-dev I see that `create-delete` sync policy has been added. I added one e2e and one unit test for this new use case

speedfl commented:
Integration test are failing because of `TestSyncOptionValidateFalse` which also fails on master.

Pending to have it solved

![image](https://user-images.githubusercontent.com/23298035/208299798-634cbdbb-4bc6-45eb-8365-ed20fe7e8485.png)

crenshaw-dev commented:
@speedfl fails on master locally or in CI? Could be an environment issue.

speedfl commented:
@crenshaw-dev this morning everything worked. The CI was green.

crenshaw-dev commented:
@speedfl I'm not going to have review time for this in the 2.7 cycle. I'm asking Ishita for help, maybe she can give it a first pass.

ishitasequeira reviewed (changes requested):
@speedfl Added my comments.

speedfl commented on pkg/apis/application/v1alpha1/applicationset_types.go:
```diff
@@ -82,11 +82,38 @@ type ApplicationMatchExpression struct {
 type ApplicationSetSyncPolicy struct {
 	// PreserveResourcesOnDeletion will preserve resources on deletion. If PreserveResourcesOnDeletion is set to true, these Applications will not be deleted.
 	PreserveResourcesOnDeletion bool `json:"preserveResourcesOnDeletion,omitempty" protobuf:"bytes,1,name=syncPolicy"`
+	// ApplicationsSync represents the policy applied on the generated applications. Possible values are create-only, create-update, sync
```
I added it in the kubebuilder comment but not in the description. Fixing it

ishitasequeira replied:
I think we missed adding `create-delete` here.

speedfl commented on applicationset/utils/policy.go:
```diff
@@ -1,55 +1,22 @@
+	"": argov1alpha1.ApplicationsSyncPolicyPolicySync,
+}
+
+func DefaultPolicy(appSetSyncPolicy *argov1alpha1.ApplicationSetSyncPolicy, defaultPolicy argov1alpha1.ApplicationsSyncPolicy, allowPolicyOverride bool) argov1alpha1.ApplicationsSyncPolicy {
+	if appSetSyncPolicy == nil || appSetSyncPolicy.ApplicationsSync == nil || !allowPolicyOverride {
+		return defaultPolicy
+	}
+	return *appSetSyncPolicy.ApplicationsSync
```
The name of the method or parameter may be misleading I guess. It does not default on the SyncPolicy `sync` but on the controller policy.

Changing the parameter from `defaultPolicy` to `controllerPolicy`

ishitasequeira replied:
Wondering if we want to change this to `controllerPolicy` as discussed in the comments.

speedfl commented on applicationset/utils/policy.go:
```diff
@@ -1,55 +1,22 @@
+	"create-update": argov1alpha1.ApplicationsSyncPolicyPolicyCreateUpdate,
+	"create-delete": argov1alpha1.ApplicationsSyncPolicyPolicyCreateDelete,
+	"sync":          argov1alpha1.ApplicationsSyncPolicyPolicySync,
+	// Default is "sync"
+	"": argov1alpha1.ApplicationsSyncPolicyPolicySync,
```
So you mean doing this in the applicationset_controller cmd

https://github.com/argoproj/argo-cd/pull/11462/files#diff-1ee9bd883626e217e79b2537de12484dd8f8e5a0d61cc792b56a86a9d9576b5b

```golang
var policyObj appv1alpha1.ApplicationsSyncPolicy
if policy == "" {
	policyObj = appv1alpha1.ApplicationsSyncPolicySync
} else {
	var exists bool
	policyObj, exists = utils.Policies[policy]
	if !exists {
		log.Info("Policy value can be: sync, create-only, create-update, create-delete. Default value: sync")
		os.Exit(1)
	}
}
```

I cannot force it in the cmd default value as the default value of `ARGOCD_APPLICATIONSET_CONTROLLER_ALLOW_POLICY_OVERRIDE` is based on the value `""` for the policy

ishitasequeira reviewed (changes requested):
I tested the PR and it works as expected. @speedfl could you rebase the PR and resolve conflicts?

ishitasequeira commented:
@speedfl the lint error is fixed. Let's rebase this PR.

speedfl commented:
Hey @ishitasequeira  I'll work on it tomorrow morning ! 🙂

speedfl commented:
Hey @ishitasequeira do you know why I am geting error on `TestNamespacedGetLogsAllowSwitchOnNS` while I did not touched it ?

ishitasequeira commented:
@speedfl AFAIK, It is a flaky test.. The solution for it is to re-trigger the test for now..

speedfl commented:
Another test fail. Is there a way to relaunch the actions without doing a commit ?

speedfl commented:
I run the test locally and it is working :/
[TestNamespacedGetLogsAllowSwitchOnNS.log](https://github.com/argoproj/argo-cd/files/11382044/TestNamespacedGetLogsAllowSwitchOnNS.log)
[TestNamespacedGetLogsAllowSwitchOnNS.run.log](https://github.com/argoproj/argo-cd/files/11382045/TestNamespacedGetLogsAllowSwitchOnNS.run.log)

speedfl commented:
@ishitasequeira seems working now :)

speedfl commented:
@ishitasequeira can we try to merge it now everything is green ?

ishitasequeira commented on pkg/apis/application/v1alpha1/applicationset_types.go:
```diff
@@ -82,11 +82,38 @@ type ApplicationMatchExpression struct {
+
+//sync / create-only / create-update / create-delete
+const (
+	ApplicationsSyncPolicyPolicyCreateOnly   ApplicationsSyncPolicy = "create-only"
+	ApplicationsSyncPolicyPolicyCreateUpdate ApplicationsSyncPolicy = "create-update"
+	ApplicationsSyncPolicyPolicyCreateDelete ApplicationsSyncPolicy = "create-delete"
+	ApplicationsSyncPolicyPolicySync         ApplicationsSyncPolicy = "sync"
```
change the constants to `ApplicationsSyncPolicyCreateOnly`, `ApplicationsSyncPolicyCreateUpdate` and so on..

speedfl commented:
@ishitasequeira should be good now. Sorry for that issue

speedfl commented:
All good now!

speedfl commented:
Hi @ishitasequeira any ETA for this one ?
We could set the milestones to 2.8 what do you think ?

ishitasequeira commented on cmd/argocd-applicationset-controller/commands/applicationset_controller.go:
```diff
@@ -55,6 +55,7 @@ func NewCommand() *cobra.Command {
 		namespace                    string
 		argocdRepoServer             string
 		policy                       string
+		allowPolicyOverride    bool
```
nit: formatting

ishitasequeira commented on cmd/argocd-applicationset-controller/commands/applicationset_controller.go:
```diff
@@ -90,7 +91,7 @@ func NewCommand() *cobra.Command {
 			policyObj, exists := utils.Policies[policy]
 			if !exists {
-				log.Info("Policy value can be: sync, create-only, create-update, create-delete")
+				log.Info("Policy value can be: sync, create-only, create-update, create-delete, '' (sync)")
```
nit: ` '' (sync)"` -> `default value: sync`

ishitasequeira commented on pkg/apis/application/v1alpha1/applicationset_types.go:
```diff
@@ -87,11 +87,38 @@ type ApplicationMatchExpression struct {
+// "create-only" means applications are only created. If the generator's result contains update, applications won't be updated
+// "create-update" means applications are only created/Updated. If the generator's result contains update, applications will be updated, but not deleted
+// "create-delete" means applications are only created/deleted. If the generator's result contains update, applications won't be updated, if it results in deleted applications, the applications will be deleted
+// "sync" means create/update/deleted. If the generator's result contains update, applications will be updated, if it results in deleted applications, the applications will be deleted
```
Let's add a statement saying, "If no ApplicationsSyncPolicy is defined, it defaults it to `sync`". WDYT?

speedfl replied:
Better you are right

ishitasequeira reviewed (changes requested):
Thanks @speedfl, The overall PR looks good. Just minor nits.

ishitasequeira commented:
> Hi @ishitasequeira any ETA for this one ? We could set the milestones to 2.8 what do you think ?

@speedfl I just left last few comments (nits).. Once that is addressed, we would need an approver for second review to get it merged. We should be able to make it to 2.8 milestone though..

cc: @crenshaw-dev

speedfl commented:
@ishitasequeira all done :)

ishitasequeira commented on docs/operator-manual/applicationset/Controlling-Resource-Modification.md:
```diff
@@ -14,17 +14,50 @@ To enable dry-run, add `--dryrun true` to the ApplicationSet Deployment's contai
+
+- Policy `create-only`: Prevents ApplicationSet controller from modifying or deleting Applications.
+- Policy `create-update`: Prevents ApplicationSet controller from deleting Applications. Update is allowed.
+- Policy `create-delete`: Prevents ApplicationSet controller from modifying Applications. Delete is allowed.
```
I think I missed this, should we add `sync` here as an option as well?

speedfl replied:
I added the policy sync.

ishitasequeira reviewed (approved):
LGTM!! Just added a quick documentation change that I missed out earlier.

@crenshaw-dev could you take another look at the PR?

crenshaw-dev reviewed (approved):
lfg, thanks as always @speedfl!
